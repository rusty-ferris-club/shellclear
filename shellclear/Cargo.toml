[package]
name = "shellclear"
version = "0.4.7"
edition = "2021"
authors = ["Elad-Kaplan <kaplan.elad@gmail.com>"]
readme = "README.md"
repository = "https://github.com/rusty-ferris-club/shellclear"
homepage = "https://github.com/rusty-ferris-club/shellclear"

[dependencies]
clap = { version = "3", features = ["cargo"], optional = true }
serde = "1"
serde_derive = "1"
serde_yaml = "^0.8.24"
serde_regex = "1.1.0"
env_logger = "^0.9.0"
log = "^0.4.17"
anyhow = "1"
console = "^0.15.0"
rayon = "1.5"
regex = "1"
chrono = "0.4"
strum = { version = "0.24", features = ["derive"] }
prettytable-rs = "^0.8"
requestty = "0.4.1"
exitcode = "^1.1.2"
dirs = "4.0"
lazy_static = "1.4.0"
mask-text = "0.1.2"

[dev-dependencies]
insta = { version = "1.17.2", features = ["backtrace", "filters"] }
tempdir = "0.3"

[build-dependencies]
anyhow = "1"

[features]
default = ["cli"]
# list optionals here:
cli = ["clap"]

[[bin]]
name = "shellclear"
required-features = ["cli"]
