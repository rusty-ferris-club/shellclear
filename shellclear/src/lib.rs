extern crate core;

mod emoji;
mod masker;
mod shell;
mod state;

pub mod config;
pub mod data;
pub mod dialog;
pub mod engine;
pub mod exporter;
pub use self::{
    emoji::Emojis,
    state::{init, ShellContext},
};
