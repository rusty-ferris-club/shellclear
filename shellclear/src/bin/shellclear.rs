use std::process::exit;

use anyhow::anyhow;
use console::{style, Style};

use shellclear::{config::Config, dialog, engine, init, Emojis, ShellContext};

mod cmd;

const DEFAULT_ERR_EXIT_CODE: i32 = 1;

pub const BANNER: &str = r#"
     _          _ _      _                 
 ___| |__   ___| | | ___| | ___  __ _ _ __ 
/ __| '_ \ / _ \ | |/ __| |/ _ \/ _` | '__|
\__ \ | | |  __/ | | (__| |  __/ (_| | |   
|___/_| |_|\___|_|_|\___|_|\___|\__,_|_| "#;

fn main() {
    let app = cmd::default::command()
        .subcommand(cmd::config::command())
        .subcommand(cmd::find::command())
        .subcommand(cmd::clear::command())
        .subcommand(cmd::stash::command());

    let matches = app.clone().get_matches();

    let env = env_logger::Env::default().filter_or(
        "LOG",
        matches.value_of("log").unwrap_or(log::Level::Info.as_str()),
    );
    env_logger::init_from_env(env);

    let config = Config::from(matches.value_of("config-dir"));
    // create app config to store state data
    let shells_context = match init() {
        Ok(s) => s,
        Err(e) => {
            log::error!("{}", e);
            exit(1)
        }
    };

    if matches.is_present("init-shell") {
        // In case of an error, we need to suppress the errors to make sure that when
        // new shell is open the users will not get any errors.
        match engine::PatternsEngine::with_config(&config) {
            Ok(engine) => {
                let emojis = Emojis::default();
                if let Ok(shell_commands) =
                    engine.find_history_commands_from_shell_list(&shells_context)
                {
                    let sensitive_commands = shell_commands.get_commands_with_secrets();

                    if sensitive_commands.is_empty() {
                        eprintln!(
                            "{} Your shell is clean from sensitive data!",
                            emojis.confetti
                        );
                    } else {
                        eprintln!(
                            "{} shellclear found {} sensitive commands in your shell history. run \
                             `shellclear find` to see more information",
                            emojis.alarm,
                            style(sensitive_commands.len()).red()
                        );
                    }
                }
            }
            Err(e) => {
                log::debug!("could not init engine config. err: {}", e);
            }
        };
        exit(0)
    }

    if !matches.is_present("no_banner") {
        println!(
            "{}{}",
            style(BANNER).magenta(),
            style(app.get_version().unwrap_or("")).dim()
        );
    }

    let res = match matches.subcommand() {
        None => Err(anyhow!("command not found")),
        Some(tup) => match tup {
            ("config", subcommand_matches) => cmd::config::run(subcommand_matches, &config),
            ("find", subcommand_matches) => {
                cmd::find::run(subcommand_matches, &shells_context, &config)
            }
            ("clear", subcommand_matches) => {
                cmd::clear::run(subcommand_matches, &shells_context, &config)
            }
            ("stash", subcommand_matches) => {
                cmd::stash::run(subcommand_matches, select_shell(&shells_context))
            }
            _ => unreachable!(),
        },
    };

    let exit_with = match res {
        Ok(cmd) => {
            if let Some(message) = cmd.message {
                let style = if exitcode::is_success(cmd.code) {
                    Style::new().green()
                } else {
                    Style::new().red()
                };
                eprintln!("\r\n{}\r\n", style.apply_to(message));
            }
            cmd.code
        }
        Err(e) => {
            log::debug!("{:?}", e);
            DEFAULT_ERR_EXIT_CODE
        }
    };
    exit(exit_with)
}

fn select_shell(shell_contexts: &Vec<ShellContext>) -> &ShellContext {
    if shell_contexts.len() == 1 {
        return &shell_contexts[0];
    }
    let selections = shell_contexts
        .iter()
        .map(|f| format!("{:?}     : {}", f.history.shell, f.history.path))
        .collect::<Vec<_>>();

    match dialog::select("Pick your shell", &selections) {
        Ok(selection) => &shell_contexts[selection],
        Err(e) => {
            log::debug!("prompter select err: {:?} ", e);
            exit(1)
        }
    }
}
