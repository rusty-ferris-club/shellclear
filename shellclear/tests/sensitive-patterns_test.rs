use insta::assert_debug_snapshot;
use serde_derive::Deserialize;
use shellclear::{data::Detection, engine::SENSITIVE_COMMANDS};

#[derive(Debug, Deserialize, Clone)]
struct TestSensitivePatterns {
    pub name: String,
    pub test: String,
    pub expected: String,
}

#[allow(dead_code)]
#[derive(Debug)]
struct TestPatternResult {
    name: String,
    test: String,
    detect: String,
    expected: String,
}

#[test]
fn can_detect_regex_sensitive_patterns() {
    let patterns: Vec<Detection> = serde_yaml::from_str(SENSITIVE_COMMANDS).unwrap();

    for pattern in &patterns {
        let pattern_id = pattern.id.to_lowercase().replace(" ", "_");
        let pattern_file = format!("tests/suites_sensitive_patterns/{}.yaml", pattern_id);
        let f = match std::fs::File::open(pattern_file) {
            Ok(f) => f,
            Err(_e) => {
                println!("not found tests for {}", pattern.id);
                return assert!(false);
            }
        };
        let tests: Vec<TestSensitivePatterns> = serde_yaml::from_reader(f).unwrap();
        let mut results = Vec::new();
        for test in tests {
            let caps = pattern.test.captures(&test.test).unwrap();
            results.push(TestPatternResult {
                name: test.name,
                test: test.test.clone(),
                detect: caps
                    .get(pattern.secret_group as usize)
                    .map_or("", |m| m.as_str())
                    .to_string(),
                expected: test.expected,
            })
        }
        assert_debug_snapshot!(pattern_id, results);
    }
}
