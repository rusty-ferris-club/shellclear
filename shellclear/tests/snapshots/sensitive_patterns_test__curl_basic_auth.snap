---
source: shellclear/tests/sensitive-patterns_test.rs
expression: results
---
[
    TestPatternResult {
        name: "detect curl -H Authorization Basic [base64_encoded_login]",
        test: "curl -X POST -H 'Authorization: Basic YmFzZTY0OnBhc3N3b3JkCg==' https://priv/",
        detect: "YmFzZTY0OnBhc3N3b3JkCg==",
        expected: "YmFzZTY0OnBhc3N3b3JkCg==",
    },
    TestPatternResult {
        name: "detect curl header authorization basic with reading from a file",
        test: "curl --header 'authorization: basic `cat /tmp/secret`' https://priv/",
        detect: "`cat /tmp/secret`",
        expected: "`cat /tmp/secret`",
    },
    TestPatternResult {
        name: "detect curl header authorization basic with reading from a file alternative method",
        test: "curl --header 'authorization: basic $(cat /tmp/secret)' https://priv/",
        detect: "$(cat /tmp/secret)",
        expected: "$(cat /tmp/secret)",
    },
    TestPatternResult {
        name: "detect curl -H authorization with secret environment variable",
        test: "curl -H 'Authorization: Basic $SECRET_VAR' https://priv/",
        detect: "$SECRET_VAR",
        expected: "$SECRET_VAR",
    },
]
